std-bits/table/src/main/java/org/enso/table/data/column/builder/InferringDoubleBuilder.java

```diff
@@ -0,0 +1,191 @@
+package org.enso.table.data.column.builder;
+
+import org.enso.base.polyglot.NumericConverter;
+import org.enso.table.data.column.storage.Storage;
+import org.enso.table.error.ValueTypeMismatchException;
+
+import java.math.BigInteger;
+import java.util.BitSet;
+
+/**
+ * A double builder variant that preserves types and can be retyped to Mixed.
+ */
+public class InferringDoubleBuilder extends DoubleBuilder {
+  /**
+   * Converts the provided LongBuilder to a DoubleBuilder.
+   *
+   * <p>The original LongBuilder becomes invalidated after this operation and should no longer be
+   * used.
+   */
+  static InferringDoubleBuilder retypeFromLongBuilder(LongBuilder longBuilder) {
+    int currentSize = longBuilder.currentSize;
+    InferringDoubleBuilder newBuilder =
+        new InferringDoubleBuilder(longBuilder.isMissing, longBuilder.data, currentSize);
+
+    // Invalidate the old builder.
+    longBuilder.data = null;
+    longBuilder.isMissing = null;
+    longBuilder.currentSize = -1;
+
+    // Assume all longs will be compacted.
+    newBuilder.isLongCompactedAsDouble.set(0, currentSize, true);
+
+    // Translate the data in-place to avoid unnecessary allocations.
+    for (int i = 0; i < currentSize; i++) {
+      if (!newBuilder.isMissing.get(i)) {
+        long currentIntegerValue = newBuilder.data[i];
+        double convertedFloatValue = (double) currentIntegerValue;
+        boolean isLossy = currentIntegerValue != (long) convertedFloatValue;
+        if (isLossy) {
+          // Save it raw for recovery.
+          newBuilder.setRaw(i, currentIntegerValue);
+          newBuilder.reportPrecisionLoss(currentIntegerValue, convertedFloatValue);
+          // Unmark the long that did not fit:
+          newBuilder.isLongCompactedAsDouble.set(i, false);
+        }
+
+        newBuilder.data[i] = Double.doubleToRawLongBits(convertedFloatValue);
+      }
+    }
+
+    return newBuilder;
+  }
+
+  InferringDoubleBuilder(BitSet isMissing, long[] doubleData, int currentSize) {
+    super(isMissing, doubleData, currentSize);
+    rawData = null;
+    isLongCompactedAsDouble = new BitSet();
+  }
+
+  /**
+   * Stores the raw data as passed to append, in order to be able to reconstruct the original values when retyping to
+   * mixed.
+   */
+  private Object[] rawData;
+
+  /**
+   * Specifies at which indices we encountered integers.
+   * <p>
+   * This is used for reconstructing the original values when retyping to mixed. Integers that are small enough can be
+   * reconstructed from their double values, so we do not need to store them as `rawData`, instead we only mark that
+   * they need to be converted back into integers when retyping. This allows us to completely avoid allocating the
+   * `rawData` array for most practical scenarios when the integers are not too large. The cost of allocating this
+   * BitSet should be significantly lower than allocating the `rawData` array.
+   */
+  private final BitSet isLongCompactedAsDouble;
```

Comment:
The `rawData` allows for reconstructing original values with original types when retyping back to Mixed.

It has to be `Object`/`Number` because we can get either `Long` or `BigInteger`, so an unboxed array wouldn't suffice.

---

The `isLongCompactedAsDouble` is an optimization that allows us to avoid allocating the potentially large `rawData` array in case all encountered `long` values are not too large - in this case they fit into `double` without losing precision, so we can reconstruct them without the helper array.

The additional logic does make the thing slower in case of encountering lots of big integer values, but it improves the 'most common' (only floats, or only floats and relatively small integers) case significantly.

The measurements are as follows:
![image](https://github.com/enso-org/enso/assets/1436948/e79758f7-3546-4f1c-ac9d-d4434edf175d)

Reply:
We can see that with the common path optimization our results in `Auto` mode are basically as fast as the old (incorrect) implementation - whereas the fixed implementation _without_ the optimization is even 2x slower (see `Floats_type_Auto` for 2x slower, `Integers_type_Auto` is just slightly slower).

In the case of mixed numbers - integers and floats; if the integers are relatively small (e.g. less than 2^50 or so - so a huge range anyway) the optimized case behaves roughly as fast as the old incorrect implementation vs the unoptimized one which is almost 2x slower (see `Numeric_Mix_Small_type_Auto`).

Finally, the last case - `Numeric_Mix_Large_type_Auto` - the case where we encounter `BigInteger` values and `long` values that cannot be exactly represented in `double`. This is the case where the 'optimized' approach can be as much as 2x slower than the 'simple' fixed solution. This is however the rarest occurrence - we don't expect to get such big numbers often. Weirdly, the 'old' implementation also had this relatively slower speed (I'm not sure why).

---

In case we know the expected type `*_type_Float`, performance of all approaches is comparable - this path is largely unaffected by these changes.

----

I think the optimization is beneficial, because it allows our most common paths to run much faster, at the cost of complicating rarely used paths.